src/dataflow/src/sink/tail.rs

```diff
@@ -30,22 +35,37 @@ pub fn tail<G>(
     let mut tx = block_on(connector.tx.connect()).expect("tail transmitter failed");
     let mut packer = RowPacker::new();
     stream.sink(Pipeline, &format!("tail-{}", id), move |input| {
-        input.for_each(|_, rows| {
-            let mut results: Vec<Row> = Vec::new();
-            for (row, time, diff) in rows.iter() {
-                let should_emit = if connector.strict {
-                    connector.frontier.less_than(time)
-                } else {
-                    connector.frontier.less_equal(time)
-                };
-                if should_emit {
-                    packer.push(Datum::Decimal(Significand::new(i128::from(*time))));
-                    packer.push(Datum::Int64(i64::cast_from(*diff)));
-                    packer.extend_by_row(row);
-                    results.push(packer.finish_and_reuse());
+        input.for_each(|_, batches| {
+            let mut results = vec![];
+            for batch in batches.iter() {
+                let mut cursor = batch.cursor();
+                while cursor.key_valid(&batch) {
+                    while cursor.val_valid(&batch) {
+                        let row = cursor.val(&batch);
+                        cursor.map_times(&batch, |time, diff| {
+                            let should_emit = if connector.strict {
+                                connector.frontier.less_than(time)
+                            } else {
+                                connector.frontier.less_equal(time)
+                            };
+                            if should_emit {
+                                packer.push(Datum::Decimal(Significand::new(i128::from(*time))));
+                                packer.push(Datum::Int64(i64::cast_from(*diff)));
+                                packer.extend_by_row(row);
+                                // Add the unpacked timestamp so we can sort by them later.
+                                results.push((*time, packer.finish_and_reuse()));
+                            }
+                        });
+                        cursor.step_val(&batch);
+                    }
+                    cursor.step_key(&batch);
                 }
             }
 
+            // Sort results by time and convert to Vec<Row>.
+            results.sort_unstable_by_key(|(time, _)| *time);
```

Comment:
Minor, but do you want `_unstable` in here? It makes what would have been a deterministic output potentially non-deterministic. If you have a preference, no worries, but these often bite us in surprising ways.

Reply:
Would it be deterministic even with the stable sort? The determinism then would have to come from a batch's cursor always producing the same rows in the same order. Does the cursor do that?

I'm ok with the unstable sort (and thus non-deterministic output) because the only thing our docs guarantee is that the timestamps don't go down. SQL already has prior art for giving rows back in whatever order it wants unless specified. I'm ok leaving it as is until a user has good reason for it. Or if cursors happen to have that guarantee already and it's easy to add (i.e., remove `_unstable`) then I'm ok adding it.

Reply:
> Does the cursor do that?

Yup, it does.

I was thinking that determinism might be good for testing, among other things. It should just be removing the `_unstable`, and you get a different performance profile (_unstable is pattern-defeating quicksort, stable is mergesort).

Reply:
what if batches are cut at different time boundaries across restarts? I'm having a hard time understanding how that is guaranteed if two batches that cover a given timestamp happen to have different sets of keys because they cover different time ranges?

Reply:
As long as all `Row` objects at a single timestamp are produced in the same order this will be deterministic. That would happen if, uh, the arrangement's key (or whatever?) is the Row? Then time boundaries won't matter. Is that what's happening?

Reply:
Yes I'm convinced. Rows are the value in this arrangement (only 1 key) and I suspect we'll see rows in order, which means that the relative order of rows at a given timestamp will be stable even if the range of timestamps changes. thanks!